File: CS/XPCustomBindingExample/Controllers/OrdersController.cs
using System;
using System.Collections;
using System.Web.Mvc;
using DevExpress.Data;
using DevExpress.Web.Mvc;
using DevExpress.Xpo.Metadata;
using XPCustomBindingExample.XPO;

namespace XPCustomBindingExample.Controllers {
    public class OrdersController :Controller {
        public ActionResult Index() {
            return View();
        }

        public ActionResult GridViewPart() {
            GridViewModel viewModel = GridViewExtension.GetViewModel("OrdersGridView");
            if(viewModel == null)
                viewModel = CreateGridViewModel();
            return OrdersCustomBindingCore(viewModel);
        }

        public ActionResult OrdersPagingAction(GridViewPagerState pager) {
            GridViewModel viewModel = GridViewExtension.GetViewModel("OrdersGridView");
            viewModel.ApplyPagingState(pager);
            return OrdersCustomBindingCore(viewModel);
        }

        public ActionResult OrdersSortingAction(GridViewColumnState column, bool reset) {
            GridViewModel viewModel = GridViewExtension.GetViewModel("OrdersGridView");
            viewModel.ApplySortingState(column, reset);
            return OrdersCustomBindingCore(viewModel);
        }

        public ActionResult OrdersFilteringAction(GridViewFilteringState filteringState) {
            GridViewModel viewModel = GridViewExtension.GetViewModel("OrdersGridView");
            viewModel.ApplyFilteringState(filteringState);
            return OrdersCustomBindingCore(viewModel);
        }

        public ActionResult OrdersGroupingAction(GridViewColumnState column) {
            GridViewModel viewModel = GridViewExtension.GetViewModel("OrdersGridView");
            viewModel.ApplyGroupingState(column);
            return OrdersCustomBindingCore(viewModel);
        }

        private static GridViewModel CreateGridViewModel() {
            GridViewModel viewModel = new GridViewModel();
            viewModel.KeyFieldName = "ID";
            viewModel.Columns.Add("ProductName");
            viewModel.Columns.Add("CustomerName");
            GridViewColumnState orderColumn = viewModel.Columns.Add("OrderDate");
            orderColumn.SortOrder = ColumnSortOrder.Ascending;
            orderColumn.SortIndex = 0;
            viewModel.Columns.Add("Freight");

            viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = "Freight", SummaryType = SummaryItemType.Average });
            viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = "OrderDate", SummaryType = SummaryItemType.Max });
            viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = "OrderDate", SummaryType = SummaryItemType.Min });
            viewModel.TotalSummary.Add(new GridViewSummaryItemState() { FieldName = "ProductName", SummaryType = SummaryItemType.Count });
            GridViewSummaryItemState groupSummary = new GridViewSummaryItemState();
            groupSummary.SummaryType = SummaryItemType.Sum;
            groupSummary.FieldName = "Freight";

            return viewModel;
        }

        private PartialViewResult OrdersCustomBindingCore(GridViewModel viewModel) {
            XPDictionary databaseSchema = DatabaseSchemaHelper.GetDefaultDatabaseSchema();
            DevExpress.Xpo.Session session = XpoHelper.GetNewSession("DefaultConnection", databaseSchema);
            XPClassInfo classInfo = session.GetClassInfo("", "Order");
            XpoBindingHandlers bindingHandlers = new XpoBindingHandlers(session, classInfo);
            viewModel.ProcessCustomBinding(
                bindingHandlers.GetDataRowCount,
                bindingHandlers.GetData,
                bindingHandlers.GetSummaryValues,
                bindingHandlers.GetGroupingInfo,
                bindingHandlers.GetUniqueHeaderFilterValues
                );
            return PartialView("GridViewPartial", viewModel);
        }
    }
}

File: VB/XPCustomBindingExample/Controllers/OrdersController.vb
Imports System
Imports System.Collections
Imports System.Web.Mvc
Imports DevExpress.Data
Imports DevExpress.Web.Mvc
Imports DevExpress.Xpo.Metadata
Imports XPCustomBindingExample.XPO

Namespace XPCustomBindingExample.Controllers
	Public Class OrdersController
		Inherits Controller

		Public Function Index() As ActionResult
			Return View()
		End Function

		Public Function GridViewPart() As ActionResult
			Dim viewModel As GridViewModel = GridViewExtension.GetViewModel("OrdersGridView")
			If viewModel Is Nothing Then
				viewModel = CreateGridViewModel()
			End If
			Return OrdersCustomBindingCore(viewModel)
		End Function

		Public Function OrdersPagingAction(ByVal pager As GridViewPagerState) As ActionResult
			Dim viewModel As GridViewModel = GridViewExtension.GetViewModel("OrdersGridView")
			viewModel.ApplyPagingState(pager)
			Return OrdersCustomBindingCore(viewModel)
		End Function

		Public Function OrdersSortingAction(ByVal column As GridViewColumnState, ByVal reset As Boolean) As ActionResult
			Dim viewModel As GridViewModel = GridViewExtension.GetViewModel("OrdersGridView")
			viewModel.ApplySortingState(column, reset)
			Return OrdersCustomBindingCore(viewModel)
		End Function

		Public Function OrdersFilteringAction(ByVal filteringState As GridViewFilteringState) As ActionResult
			Dim viewModel As GridViewModel = GridViewExtension.GetViewModel("OrdersGridView")
			viewModel.ApplyFilteringState(filteringState)
			Return OrdersCustomBindingCore(viewModel)
		End Function

		Public Function OrdersGroupingAction(ByVal column As GridViewColumnState) As ActionResult
			Dim viewModel As GridViewModel = GridViewExtension.GetViewModel("OrdersGridView")
			viewModel.ApplyGroupingState(column)
			Return OrdersCustomBindingCore(viewModel)
		End Function

		Private Shared Function CreateGridViewModel() As GridViewModel
			Dim viewModel As New GridViewModel()
			viewModel.KeyFieldName = "ID"
			viewModel.Columns.Add("ProductName")
			viewModel.Columns.Add("CustomerName")
			Dim orderColumn As GridViewColumnState = viewModel.Columns.Add("OrderDate")
			orderColumn.SortOrder = ColumnSortOrder.Ascending
			orderColumn.SortIndex = 0
			viewModel.Columns.Add("Freight")

			viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = "Freight", .SummaryType = SummaryItemType.Average})
			viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = "OrderDate", .SummaryType = SummaryItemType.Max})
			viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = "OrderDate", .SummaryType = SummaryItemType.Min})
			viewModel.TotalSummary.Add(New GridViewSummaryItemState() With {.FieldName = "ProductName", .SummaryType = SummaryItemType.Count})
			Dim groupSummary As New GridViewSummaryItemState()
			groupSummary.SummaryType = SummaryItemType.Sum
			groupSummary.FieldName = "Freight"

			Return viewModel
		End Function

		Private Function OrdersCustomBindingCore(ByVal viewModel As GridViewModel) As PartialViewResult
			Dim databaseSchema As XPDictionary = DatabaseSchemaHelper.GetDefaultDatabaseSchema()
            Dim xpoSession As DevExpress.Xpo.Session = XpoHelper.GetNewSession("DefaultConnection", databaseSchema)
            Dim classInfo As XPClassInfo = xpoSession.GetClassInfo("", "Order")
            Dim bindingHandlers As New XpoBindingHandlers(xpoSession, classInfo)
            viewModel.ProcessCustomBinding(AddressOf bindingHandlers.GetDataRowCount, AddressOf bindingHandlers.GetData, AddressOf bindingHandlers.GetSummaryValues, AddressOf bindingHandlers.GetGroupingInfo, AddressOf bindingHandlers.GetUniqueHeaderFilterValues)
			Return PartialView("GridViewPartial", viewModel)
		End Function
	End Class
End Namespace